networkx/drawing/layout.py

```diff
@@ -1121,6 +1122,235 @@ def multipartite_layout(G, subset_key="subset", align="vertical", scale=1, cente
+        max_jitter = 10
+        min_speed_efficiency = 0.05
+
+        other = min(max_jitter, opt_jitter * traction / n**2)
+        jitter = jitter_tolerance * max(min_jitter, other)
+
+        if swing / traction > 2.0:
+            if speed_efficiency > min_speed_efficiency:
+                speed_efficiency *= 0.5
+            jitter = max(jitter, jitter_tolerance)
+        if swing == 0:
+            target_speed = np.inf
+        else:
+            target_speed = jitter * speed_efficiency * traction / swing
+
+        if swing > jitter * traction:
+            if speed_efficiency > min_speed_efficiency:
+                speed_efficiency *= 0.7
+        elif speed < 1000:
+            speed_efficiency *= 1.3
+
+        max_rise = 0.5
+        speed = speed + min(target_speed - speed, max_rise * speed)
+        return speed, speed_efficiency
+
+    # parse optional pos positions
+    if pos is None:
+        pos = nx.random_layout(G, dim=dim)
+    else:
+        dim = len(next(iter(pos.values())))
+
+    # Scale the position of the node proportional to the pos
+    max_pos_range = 1
+    min_pos_range = 0
+    # check if we have a valid pos else just return (empty graph)
+    if pos:
+        max_pos_range = np.array([max(i) for i in pos.values()]).max()
+        min_pos_range = np.array([min(i) for i in pos.values()]).min()
+    else:
+        return pos
+
+    # center locs around 0
+    pos_arr = (np.random.rand(len(G), dim) - 0.5) * (max_pos_range - min_pos_range)
+
+    masses = np.zeros(len(G))
+    sizes = np.zeros(len(G))
+    for idx, node in enumerate(G.nodes()):
+        if node in pos:
+            pos_arr[idx] = pos[node].copy()
+        masses[idx] = G.nodes[node].get(mass, G.degree(node) + 1)
+        sizes[idx] = G.nodes[node].get(size, 1)
+
+    n = G.number_of_nodes()
+    gravities = np.zeros((n, dim))
+    attraction = np.zeros((n, dim))
+    repulsion = np.zeros((n, dim))
+    A = nx.to_numpy_array(G, weight=weight)
+
+    speed = 1
+    speed_efficiency = 1
+    swing = 1
+    traction = 1
+    for _ in range(n_iter):
+        # compute pairwise difference
+        diff = pos_arr[:, None] - pos_arr[None]
+        # compute pairwise distance
+        distance = np.linalg.norm(diff, axis=-1)
+
+        # linear attraction
+        if linlog:
+            attraction = -np.log(1 + distance) / distance
+            np.fill_diagonal(attraction, 0)
+            attraction = np.einsum("ij, ij -> ij", attraction, A)
+            attraction = np.einsum("ijk, ij -> ik", diff, attraction)
+
+        else:
+            attraction = -np.einsum("ijk, ij -> ik", diff, A)
+
+        if distributed_action:
+            attraction /= masses[:, None]
+
+        # repulsion
+        tmp = masses[:, None] @ masses[None]
+        if adjust_sizes:
+            distance += -sizes[:, None] - sizes[None]
+            # multiply negative distance * 100 (squared below)
+            distance[distance < 0] = 1 / 10
```

Comment:
This comment about treatment of negative distances (inside the adjust_sizes logic) doesn't make sense to me. Is it out of date? Should it be removed?

Reply:
Can't remember why this is here. I ported the code from the java implementation. With some quick testing I didn't spot a noticeable difference. I will remove it for now.